### ISSUE TITLE
Getting some weird errors in fabricate

### ISSUE THREAD

**Issue by lukesonnet** [DeclareDesign/fabricatr#92]
Holds for CRAN or github version AFAICT.
```
library(fabricatr)
dat <- fabricate(
  N = 40,
  y = rpois(N, lambda = 4),
  z = rbinom(N, 1, prob = 0.4),
  d = rbinom(N, 1, prob = 0.8)
)
#> Error in rbinom(N, 1, prob = 0.8) : object 'N' not found
```

But if I rename `d`...
```
library(fabricatr)
dat <- fabricate(
  N = 40,
  y = rpois(N, lambda = 4),
  z = rbinom(N, 1, prob = 0.4),
  x = rbinom(N, 1, prob = 0.8)
)
```

Note that there is no `d` in my environment. Happens in R and RStudio. The first call with `d` gave the below error when my env was full of junk, but I can't replicate it now:
```
Error in fabricate(N = 40, y = rpois(N, lambda = 4), z = rbinom(N, 1,  : 
  You must do exactly one of: 
1) One or more level calls, with or without existing data 
2) Import existing data and optionally, add new variables without adding a level 
3) Provide an `N` without importing data and optionally, add new variables
```

**Comment by aaronrudkin** [comment]
Oh, actually, this is less weird that what I thought. `d = ` is autocompleting to `data = ` through R's stupid-ass argument auto-completion nonsense. The error message you get depends on what all is in the argument formals; if you are creating variables other than this, you should get the "wait, you're both importing data and specifying N, why are you doing that?" error as in the bottom of the issue; if you are only creating this, it'll assume you're importing data, and complain that there's no `N` because the data import is not evaluated in an environment where `N` exists.

None of these will work:
```
fabricate(
  N = 40,
  dat = rnorm(N)
)
```

```
fabricate(
  N = 40,
  da = rnorm(N)
)
```

```
fabricate(
  N = 40,
  data = rnorm(N)
)
```

But this will:
```
fabricate(
  N = 40,
  dz = rnorm(N)
)
```

Is there a way to force R not to do the argument name autocompletion thing?

**Comment by aaronrudkin** [comment]
Note to self: I can fix this by moving the data argument past the splat and making the first argument in `fabricate()` be `.data` and linking the two. Unclear whether this hack is worth doing.

**Comment by nfultz** [comment]
Don't create a .data arg - instead, move data after the ...

Inside fabricate , test if both N and data are there or not - if niether
are there, force the evaluation of the first unnamed dot argument `i <
-which.min(names(data_arguments) == "")`

**Comment by nfultz** [comment]
duplicate of aarons ticket.

**Comment by aaronrudkin** [comment]
So as this relates to the current ecosystem, the formals become `fabricate(..., N, data, ID_label)`. The existing code checks to make sure you specify exactly 1 of N, data, or a series of level calls, and if not error.

The new code will say:
1. If you specified exactly one of those, go for it.
2. If you specify exactly 0 of those, try to process the first `...` argument as data.
3. If you specify 2 or more of those, you still screwed up.

**Comment by nfultz** [comment]
I think it should also error if `is.integer(N) && is.data.frame(..1)`

**Comment by graemeblair** [comment]
but does this mean the function is not tidy anymore, i.e. can fabricate still be put in a pipeline with this order? I thought data had to be first. pls correct me!

**Comment by nfultz** [comment]
If we handle ..1 then it should be fine. Might want to do a quick check before making a big change though.

**Comment by aaronrudkin** [comment]
Yes I would definitely not commit this change without confirming that the first argument processing works properly. I think Neal's suggestion should work and am going to plug at it now.

**Comment by aaronrudkin** [comment]
```
library(rlang)
library(dplyr)

tmp = function(..., data = NULL) {
  args = quos(...)
  first_unnamed_arg = which(names(args) == "")[1] 
  # will be NA if no args are unnamed
  
  if(is.null(data)) {
    if(is.na(first_unnamed_arg)) {
      stop("no unnamed arguments")
    }
    data = eval_tidy(args[[first_unnamed_arg]])
    args = args[-first_unnamed_arg]
  }
  
  print(data)
  print(length(args))
}

tmp("data", "other arg", "other other arg") # should be "data", 2
tmp("other arg", "other other arg", data = "data") # should be "data", 2
tmp(n1 = "other arg", n2 = "other other arg", data = "data") # should be "data", 2
tmp(n1 = "other arg", n2 = "other other arg", "data") # should be "data", 2
tmp(n1 = "data") # should error, since there's no unnamed argument
"data" %>% tmp() # should be "data", 0
```

this is a sort of minimum test code -- fabricatr's use case is a little more complicated, but it does appear to be working the way we'd expect. if someone passes the data argument it's used as is. if they don't, the first unnamed argument is considered the data argument. if there's none, something went wrong. piping works.

**Comment by aaronrudkin** [comment]
We (probably?) can't use `is.data.frame()` on the first unnamed arg because we allow passing multiple dfs in a list for double-importing, which we allow for panel/cross-classified data. But should be able to do this quickly.

**Comment by nfultz** [comment]
Level calls with out a data set would be an error, no?

**Comment by aaronrudkin** [comment]
Long story short: We specifically had a test written for `fabricate(data = NULL, add_level(N = 10, ID_label = "test"))` to allow the back-door level name injection. This is a valid call, it's an all-levels call of fabricate with the name specified in a different way. But when I added the "if you don't have a data or N, try the first unnamed argument to be data", it caused this code to error, because it tried to evaluate the `add_level` call but failed because it thought the call wasn't inside a `fabricate` call and the environment isn't sane. *exhales*

So now I have:
1. Check if the user has supplied data or N. If not, evaluate the first unnamed argument that isn't a level call, and based on the type, try to inject that as data or N. If there isn't an unnamed non-level argument, do nothing, If the data type is wrong, do nothing. 

2. Subject to whatever happened in 1, confirm the user has done exactly one of: a) all level calls, with or without data; b) data and optionally variables, c) N and optionally variables

Just have to write a few tests and we should be good. 🤞

**Comment by graemeblair** [comment]
awesome, thanks aaron!
